File: ios/Sources/DKMadsSSPSDK/UI/DKMadsHtml5PackageFit.swift
import CoreGraphics
import Foundation

/// Outer scale/letterbox for hosted HTML5 packages in fullscreen (interstitial / app open).
/// Does not mutate creative DOM — sizes the WebView to package pixels, then scales to fit.
enum DKMadsHtml5PackageFit {
    /// Bid / IAB package size; interstitial HTML5 defaults to 320×480 when missing.
    static func packageSize(for ad: Ad, defaultWidth: CGFloat = 320, defaultHeight: CGFloat = 480) -> CGSize {
        let w = ad.slotW > 0 ? CGFloat(ad.slotW) : (ad.width > 0 ? CGFloat(ad.width) : defaultWidth)
        let h = ad.slotH > 0 ? CGFloat(ad.slotH) : (ad.height > 0 ? CGFloat(ad.height) : defaultHeight)
        return CGSize(width: max(1, w), height: max(1, h))
    }

    /// Uniform scale so `package` fits inside `container` (contain / letterbox).
    static func containScale(package: CGSize, container: CGSize) -> CGFloat {
        guard package.width > 0, package.height > 0, container.width > 0, container.height > 0 else { return 1 }
        return min(container.width / package.width, container.height / package.height)
    }
}

File: android/DKMadsHtml5PackageFit.kt
package com.dkmads.ssp

/**
 * Outer scale/letterbox for hosted HTML5 packages in fullscreen (interstitial / app open).
 * Does not mutate creative DOM — sizes the WebView to package pixels, then scales to fit.
 */
object DKMadsHtml5PackageFit {
    /** Bid / IAB package size; interstitial HTML5 defaults to 320×480 when missing. */
    fun packageSize(ad: Ad, defaultWidth: Int = 320, defaultHeight: Int = 480): Pair<Int, Int> {
        val w = when {
            ad.slotW > 0 -> ad.slotW
            ad.width > 0 -> ad.width
            else -> defaultWidth
        }
        val h = when {
            ad.slotH > 0 -> ad.slotH
            ad.height > 0 -> ad.height
            else -> defaultHeight
        }
        return w.coerceAtLeast(1) to h.coerceAtLeast(1)
    }

    /** Uniform scale so [packageW]×[packageH] fits inside the container (contain / letterbox). */
    fun containScale(packageW: Float, packageH: Float, containerW: Float, containerH: Float): Float {
        if (packageW <= 0f || packageH <= 0f || containerW <= 0f || containerH <= 0f) return 1f
        return minOf(containerW / packageW, containerH / packageH)
    }
}
